machine/machine_core/ssh_connection.py

```diff
@@ -306,70 +301,26 @@ def execute(self, command=None, script=None, input=None, environment={},
-            assert not input, "input not supported to script"
-            cmd += ["sh", "-s"]
-            if self.verbose:
-                cmd += ["-x"]
-            input = env_script
-            input += script
-            command = "<script>"
+            # use shlex.join() once Python 3.8 is available everywhere
+            cmd.append(' '.join(shlex.quote(arg) for arg in command))
+            if not quiet:
+                self.message("+", *command)
         command_line = ssh_env + default_ssh_params + additional_ssh_params + env_command + cmd
 
         with timeoutlib.Timeout(seconds=timeout, error_message="Timed out on '%s'" % command, machine=self):
-            if stdout:
-                subprocess.call(command_line, stdout=stdout)
-                return
-
-            output = ""
-            proc = subprocess.Popen(command_line, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-            stdin_fd = proc.stdin.fileno()
-            stdout_fd = proc.stdout.fileno()
-            stderr_fd = proc.stderr.fileno()
-            rset = [stdout_fd, stderr_fd]
-            wset = [stdin_fd]
-            while len(rset) > 0 or len(wset) > 0:
-                ret = select.select(rset, wset, [], 10)
-                for fd in ret[0]:
-                    if fd == stdout_fd:
-                        data = os.read(fd, 1024)
-                        if not data:
-                            rset.remove(stdout_fd)
-                            proc.stdout.close()
-                        else:
-                            if self.verbose:
-                                write_all(sys.__stdout__.fileno(), data)
-                            output += data.decode('utf-8', 'replace')
-                    elif fd == stderr_fd:
-                        data = os.read(fd, 1024)
-                        if not data:
-                            rset.remove(stderr_fd)
-                            proc.stderr.close()
-                        elif not quiet or self.verbose:
-                            write_all(sys.__stderr__.fileno(), data)
-                for fd in ret[1]:
-                    if fd == stdin_fd:
-                        if input:
-                            num = os.write(fd, input.encode('utf-8'))
-                            input = input[num:]
-                        if not input:
-                            wset.remove(stdin_fd)
-                            proc.stdin.close()
-            proc.wait()
-
-        if proc.returncode != 0 and check:
-            raise subprocess.CalledProcessError(proc.returncode, command, output=output)
-        return output
+            res = subprocess.run(command_line,
+                                 stdin=None if input is not None else subprocess.DEVNULL,
+                                 input=input,
+                                 stdout=stdout or subprocess.PIPE,
+                                 check=check,
+                                 text=True)
```

Comment:
Using `text=true` would have been nice, but we have [some tests](https://cockpit-logs.us-east-1.linodeobjects.com/pull-3709-20220811-142551-14c3d745-rhel-8-7-cockpit-project-cockpit/log.html) that produce binary garbage. So I'm going to bring back the original `decode('utf-8', 'replace')`.

Reply:
Oh, `subprocess.run()` has an `errors=` kwarg, nice! (they really thought about everything..)

Reply:
Bah -- [this test](https://cockpit-logs.us-east-1.linodeobjects.com/pull-3709-20220812-053048-dd254213-debian-stable-cockpit-project-cockpit/log.html#19-2) is too strict and expects `\r\n`, which `text=True` eats up. So, back to manual .decode() after all..